ml-agents-envs/mlagents/envs/base_env.py

```diff
@@ -0,0 +1,305 @@
+"""
+Python Environment API for the ML-Agents toolkit
+The aim of this API is to expose groups of similar Agents evolving in Unity
+to perform reinforcement learning on.
+There can be multiple groups of similar Agents (same observations and actions
+spaces) in the simulation. These groups are identified by a agent_group that
+corresponds to a single group of Agents in the simulation.
+For performance reasons, the data of each group of agents is processed in a
+batched manner. When retrieving the state of a group of Agents, said state
+contains the data for the whole group. Agents in these groups are identified
+by a unique int identifier that allows tracking of Agents across simulation
+steps. Note that there is no guarantee that the number or order of the Agents
+in the state will be consistent across simulation steps.
+A simulation steps corresponds to moving the simulation forward until at least
+one agent in the simulation sends its observations to Python again. Since
+Agents can request decisions at different frequencies, a simulation step does
+not necessarily correspond to a fixed simulation time increment.
+Changes from ML-Agents v0.11 :
+ - Step now takes no arguments and returns None.
+ - The data received from a step is no longer a dictionary of NamedTuple, the
+state data must now be retrieved manually from the environment object.
+ - Reset does no longer take any input arguments.
+ - Modifying properties of the environment besides actions is handled in
+SideChannels.
+"""
+
+from abc import ABC, abstractmethod
+from typing import List, NamedTuple, Tuple, Optional, Union, Dict
+import numpy as np
+from enum import Enum
+
+
+class StepResult(NamedTuple):
+    """
+    Contains the data a single Agent collected since the last
+    simulation step.
+     - obs is a list of numpy arrays observations collected by the group of
+     agent.
+     - reward is a float. Corresponds to the rewards collected by the agent
+     since the last simulation step.
+     - done is a bool. Is true if the Agent was terminated during the last
+     simulation step.
+     - max_step is a bool. Is true if the Agent reached its maximum number of
+     steps during the last simulation step.
+     - agent_id is an int and an unique identifier for the corresponding Agent.
+     - action_mask is an optional list of one dimensional array of booleans.
+     Only available in multi-discrete action space type.
+     Each array corresponds to an action branch. Each array contains a mask
+     for each action of the branch. If true, the action is not available for
+     the agent during this simulation step.
+    """
+
+    obs: List[np.array]
+    reward: float
+    done: bool
+    max_step: bool
+    agent_id: int
+    action_mask: Optional[List[np.array]]
+
+
+class BatchedStepResult(object):
```

Comment:
nit: don't need `(object)` in python3
```suggestion
class BatchedStepResult:
```